### PR TITLE
Qute - escape expressions in HTML by default

### LINKED ISSUE #6155: Handle escaping in Qute
Right now, Qute doesn't handle escaping at all and doesn't provide any facility for escaping. This is an issue as that means that users will easily have XSS issues.

My inclination would be to enable escaping by default.

The escaping should be different for each content-type.

We need to check how we will properly handle tag content escaping and attribute escaping in HTML. We need to be sure our escaping works for both and a value escaped in an attribute still has the same value as previously when submitted through a form (think of `input value=""`: if we inject a value here, we want the submitted value to be exactly the same as the injected one).

### PR DESCRIPTION
**Pull request by mkouba**
- resolves #6155
- API changes:
 - introduce TemplateLocator
 - add Template#getVariant()

What this PR does?

1. For HTML templates `'`,`"`,`<`, `>` and `&` are escaped by default
2. It's possilble to use `raw` and `safe` extension methods for String to "unescape" the value; ie. `{foo.myString.raw}`
3. `RawString` is never escaped
4. `Template` now has an optional `Variant`

### CODE REVIEW

**Comment by gsmet** [comment]
I would lean towards including that one into 1.1. I'm not very happy to release a templating engine without default escaping. WDYT?

Any chance you could add some documentation about this behavior?

**Comment by mkouba** [comment]
> I would lean towards including that one into 1.1

I thought that 1.1 is released already ;-). If it's easy to backport (and I believe it should be because we did not change in Qute since 1.1 code freeze) then +1.

> Any chance you could add some documentation about this behavior?

+1, I'll mention this in the docs.

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java
Comment (gsmet) [inline extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java]: Shouldn't we apply that one for XML too?

Comment (mkouba) [reply extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java]: Yes, we could.

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java
Comment (gastaldi) [inline independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: The auto-boxing here looks like an expensive operation to be done on every character.

Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: Well, `Map#get()` is probably even more expensive. Anyway, it's an implementation detail that can be improved in the future... feel free to improve the algorithm ;-).

Comment (gastaldi) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: Yeah, I was looking at https://www.eclipse.org/collections/javadoc/10.0.0/org/eclipse/collections/impl/map/mutable/primitive/CharObjectHashMap.html. I'll wait until your PR is merged to perform some performance tests using it ;)

Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: There is also `io.netty.util.collection.CharObjectHashMap<V>`.

Comment (gastaldi) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: Yeah, I've been looking at that too initially. Maybe it's simple to just copy it over to Qute than to depend on another library. OSS FTW :)

**Comment by mkouba** [comment]
Hm, I've been thinking about the implementation and it's very likely not correct. The `ResultMapper` that performs escaping is only applied to expression values that resolve to a `String`. However, the escaping should be performed _after_ the value is converted to `String`. 

I'll need to change the default `ResultMapper` to convert everything to String first (if needed) and then escape the value. Custom `ResultMapper` can be still implemented if it's using higher `getPriority()`.

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java
Comment (FroMage) [inline independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: Watch out that this is not codepoint-compliant. Not sure if 16-bit components of 32-bit codepoints can be in the ascii plane, but we should check.

Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: I have no idea what you're talking about ;-). Chars in Java are unicode, or? I would be more than grateful if you could check whatever needs to be checked.

Comment (FroMage) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: chars in java are "half" unicode. they're 16 bits, so for any "unicode char" (which is called a "codepoint") that requires more than 16 bits they're encoded on two consecutive 16-bit chars using UTF-16 (standard encoding of all Java Strings at runtime).

Comment (FroMage) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: Thus iterating chars for comparison is rarely right: you must use the codepoint static methods on `Character` to ask if a codepoint is on 1 or 2 chars. Now, it's possible that the 2-char codepoints have both chars greater than the ascii range, and you're only escaping (5) ascii values so your code may work, but we should check.

Comment (FroMage) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Escaper.java]: Both surrogate pairs have their sign bits at 1 so they can't be mistaken for the ascii we're replacing, so we're good.

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java
Comment (FroMage) [inline extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java]: https://en.wikipedia.org/wiki/List_of_XML_and_HTML_character_entity_references#Predefined_entities_in_XML says `&apos;` is available for both xml and html.

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java
Comment (FroMage) [inline extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java]: I'm guessing in the future we'll need to turn this into an SPI for custom escaping of other variant content types. But it can wait.

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java
Comment (FroMage) [inline extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java]: Are we sure it's the last mapper? I don't think users should be able to register mappers _after_ this one, no?

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java
Comment (FroMage) [inline extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java]: We support JSON but I didn't see any escaper for JSON.

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/Variant.java
Comment (FroMage) [inline independent-projects/qute/core/src/main/java/io/quarkus/qute/Variant.java]: Note that XML can be `application/xml` in many cases, but I don't think it matters here. https://www.ietf.org/rfc/rfc2376.txt

**Review by FroMage — APPROVED** [review]
LGTM

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java
Comment (mkouba) [reply extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java]: Mappers with higher priority are tried first. So if a user wants to map some object then the priority should be > `io.quarkus.qute.WithPriority.DEFAULT_PRIORITY`.

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java
Comment (mkouba) [reply extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/VariantTemplateProducer.java]: Well, actually we don't support JSON ATM ;-). TBD

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/Variant.java
Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/Variant.java]: Yes, I know. But in our case we guess the content type from a file suffix and set always `text/xml` for `*.xml` templates.

File: extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java
Comment (mkouba) [reply extensions/qute/runtime/src/main/java/io/quarkus/qute/runtime/EngineProducer.java]: +1 for proper SPI... in the furure ;-).

File: independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java
Comment (gastaldi) [inline independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java]:
```suggestion
    public void testRawStringResolver() {
```

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateLocator.java
Comment (gastaldi) [inline independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateLocator.java]: What about multiple localized templates for a given template? Should this return a List?

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
Let's get moving, we can adjust later but we need at least some escaping in 1.1.

File: independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateLocator.java
Comment (mkouba) [reply independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateLocator.java]: I don't think so. This represents a single template file.

File: independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java
Comment (mkouba) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java]: I may be blind but what's the suggested change?

Comment (gastaldi) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java]: Revolver != resolver :smile:

Comment (mkouba) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java]: Ok, it's confirmed. I'm blind...

Comment (FroMage) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java]: Most beautiful typo I've seen so far :)

Comment (mkouba) [reply independent-projects/qute/core/src/test/java/io/quarkus/qute/EscaperTest.java]: I'm very good at typos!

**Pull request merged**
